### PR TITLE
Raise user error for joint parameter mutation export

### LINKED ISSUE #122860: Do not use assert for user triggered issues
### 🐛 Describe the bug

In torch/_functorch/_aot_autograd/schemas.py, we have this code:

```
        for idx, input_info in enumerate(view_mutation_metadata.input_info):
            if input_info.mutates_data:
                # Only buffers can be mutated, not parameters
                assert idx >= len(parameters)
                mutations.append(names[idx + num_tokens])
```

Please don't write asserts like this that are enforcing invariants about a user's program. Minimally, this assert should have a message roughly what the comment above it says. Ideally, it raises a user oriented exception that indicates that something is being attempted that is not legal.

### Versions

Collecting environment information...
PyTorch version: 2.3.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.28.3
Libc version: glibc-2.35

Python version: 3.11.0rc1 (main, Aug 12 2022, 10:02:14) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-25-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             64
On-line CPU(s) list:                0-63
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen Threadripper 3970X 32-Core Processor
CPU family:                         23
Model:                              49
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          1
Stepping:                           0
Frequency boost:                    enabled
CPU max MHz:                        3700.0000
CPU min MHz:                        2200.0000
BogoMIPS:                           7400.63
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es
Virtualization:                     AMD-V
L1d cache:                          1 MiB (32 instances)
L1i cache:                          1 MiB (32 instances)
L2 cache:                           16 MiB (32 instances)
L3 cache:                           128 MiB (8 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-63
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow: Mitigation; Safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] onnx==1.15.0
[pip3] torch==2.3.0+cpu
[pip3] torchaudio==2.3.0+cpu
[pip3] torchvision==0.18.0+cpu
[pip3] triton==2.1.0
[conda] Could not collect

cc @chauhang @penguinwu @Chillee @samdow @kshitij12345 @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @bdhirsh @bobrenjc93 @aorenste @ezyang @zou3519 @janeyx99 @gmagogsfm

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186470

Joint forward/backward export cannot represent module parameters as mutable
state. Forward mutations previously flowed into signature construction and
remapping paths that assumed the target was a buffer or user input, producing
internal assertions. Backward-only mutations use separate AOT metadata, so
checking only forward mutation metadata also allowed a registered parameter to
be mutated by an in-graph copy_.

The ExportedProgram path has another wrinkle: its parameters are already lifted
when the inner AOT export runs, so AOT classifies them as user inputs. A
preserved custom-op backward could therefore mutate a parameter and reach
ExportedProgram validation as a non-functional copy_, producing an internal
SpecViolationError.

Reject registered parameter mutations in the direct AOT path and retain the
names of inputs mutated only during backward in the internal AOT graph
signature. The ExportedProgram path maps those names back to its original input
specs and raises the same user-facing RuntimeError for parameters. A defaulted
signature field preserves existing constructors, and supported forward buffer
and user-input mutations are unchanged. Rejecting the direct backward-only
parameter case is intentional: AOT can keep the copy_ in its graph, but a
module parameter is not representable as mutable state in this joint export
contract.

Carrying the existing mutation metadata avoids scanning the generated graph for
copy_ nodes and keeps parameter classification at the signature boundary where
the original input kinds are available.

Fixes #122860
Generated by my agent

Test Plan:
```
python test/export/test_experimental.py TestExperiment.test_joint_parameter_mutation_error TestExperiment.test_joint_aot_buffer_mutation_still_allowed TestExperiment.test_joint_buffer_input_mutations
python test/export/test_export.py TestDynamismExpression.test_no_grad_param_inplace
source .venv/bin/activate && lintrunner -a
```

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the failing inductor_huggingface CI job was caused by DebertaV2ForMaskedLM being loaded before its existing accuracy skip could record pass_due_to_skip; the updated PR now skips that model at HuggingFace benchmark selection time while preserving the required issue-closing metadata.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and removed the DebertaV2ForMaskedLM benchmark skip from this diff because that workaround already landed through #186248. The export RuntimeError fix remains scoped to the original issue, and the refreshed tests now cover both raw and already-decomposed ExportedProgram mutation metadata.

File: torch/_functorch/_aot_autograd/schemas.py [hunk 1]
```diff
@@ -1063,8 +1063,12 @@ def from_tracing_metadata(
                 if trace_joint:
                     # Only buffers can be mutated, not parameters
                     if idx < len(parameters):
-                        raise AssertionError(
-                            f"expected idx ({idx}) >= len(parameters) ({len(parameters)}) when tracing joint"
+                        raise RuntimeError(
```
Comment (bobrenjc93) [inline torch/_functorch/_aot_autograd/schemas.py]: This only checks forward mutation metadata. Joint capture records inputs mutated exclusively during backward in `indices_of_inputs_that_requires_grad_with_mutations_in_bw`, leaving `input_info.mutates_data` false, so a parameter mutated by a custom autograd backward bypasses this error. Reject parameter indices from that list as well and add a regression test.

Comment (jansel) [reply torch/_functorch/_aot_autograd/schemas.py]: My agent says this update addresses the feedback:

> The joint signature check now rejects parameter indices recorded in `indices_of_inputs_that_requires_grad_with_mutations_in_bw`, not just forward `mutates_data` metadata. A custom-autograd regression now mutates `p` exclusively during backward and verifies the same user-facing `RuntimeError`.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this:

> The latest update closes the remaining backward-only mutation gap in `_export_forward_backward`. A preserved custom-op backward could mutate a lifted parameter after the inner AOT export lost its outer input kind, leading to an internal `SpecViolationError` on `aten.copy_`. The AOT signature now carries the backward-mutated input names so export can remap them to the original parameter and raise the intended user-facing error. Raw and decomposed regression cases, adjacent export tests, and `lintrunner -a` pass.

**Comment by jansel** [comment]
@claude review these changes

**Review by mlazos — APPROVED** [review]

**Comment by jansel** [comment]
@bobrenjc93 can you look again
